### ISSUE TITLE
Improve performance

### ISSUE THREAD

**Issue by Keats** [Keats/tera#84]
I added some benches and performance is not that great when rendering both inheritance and macros or using lots of variable across 3 templates, a pretty common case:

```bash
test bench_rendering_basic_template            ... bench:      11,157 ns/iter (+/- 2,765)
test bench_rendering_inheritance_and_macros    ... bench:      28,078 ns/iter (+/- 13,440)
```
Note that my laptop is ~3y old so someone with a good desktop/server will beat that easily.
Results are in `ns` so it's still pretty fast but can easily be improved as I haven't cared about performance yet and there is lots of `clone()` around.

I only mention rendering performance as parsing performance is not critical since that's a one-time cost and is fast enough. But i'm obviously pro performance gains there as well

**Comment by clarcharr** [comment]
I addressed some of the clones in #113.

**Comment by clarcharr** [comment]
A few other ideas I came up with:

1. Make AST use an Arena for much better cache locality
1. Make `Node::get_children` use references instead of cloning.
1. Make `Render` use `Display` instead of copying around `String`s
1. Replace occurrences of `String` with `Cow<str>`
1. Rewrite `args` for filters accept to accept enums for known argument names instead of always using `String` keys

**Comment by Keats** [comment]
I'm ok with 1 to 4 if that makes a different. For 5, how would it work for the signature of a filter function and user-defined filters?

**Comment by clarcharr** [comment]
It would be a breaking change; I actually tried it and it didn't seem to make much of a difference in performance, so, I've removed it from the list.

**Comment by Keats** [comment]
@clarcharr are you working on any of those? I might pick some of those to try when I have time

**Comment by clarcharr** [comment]
@keats Not at the moment! Feel free

**Comment by Keats** [comment]
Got some pretty nice improvements juste reducing the number of cloning:

```bash
 name                                       control.bench ns/iter  now.bench ns/iter  diff ns/iter   diff % 
 bench_escape_html                          133                    130                          -3   -2.26% 
 bench_parsing_basic_template               12,950                 9,614                    -3,336  -25.76% 
 bench_parsing_with_inheritance_and_macros  40,578                 33,391                   -7,187  -17.71% 
 bench_rendering_basic_template             7,675                  6,540                    -1,135  -14.79% 
 bench_rendering_inheritance_and_macros     15,781                 11,110                   -4,671  -29.60% 
 bench_rendering_only_inheritance           3,496                  2,779                      -717  -20.51% 
 bench_rendering_only_macro_call            2,260                  1,437                      -823  -36.42% 
 bench_rendering_only_parent                2,195                  1,375                      -820  -37.36% 
 bench_rendering_only_variable              1,869                  1,712                      -157   -8.40% 
```
in the `next` branch
There's probably still too much cloning going on with the `Value` objects (ie currently a lookup clones the value, parsing a forloop clones the array, etc) but that's a good start. Hopefully even better performance coming soon!

**Comment by Peternator7** [comment]
I was experimenting with the rendering this past weekend. I had some success using `Cow<str>` but it felt like it made things quite a bit more complicated for relatively small performance gains.

I did see modest improvements (~10%) by having each render function write directly to the output string rather than return it's output, and that eliminates the need for `Cow<>`'s all together. It doesn't quite pass all the tests yet because the nuances of trimming the strings are a little different, but if that's an architectural change you would be interested in making, I think there is opportunity for increased performance.

**Comment by Keats** [comment]
I'm not entirely sure how that's going to work for the trimming but if you manage to figure it out, I'll take it!
Make sure you work off the `next` branch though, not `master`.

**Comment by Keats** [comment]
@Peternator7 i've released 0.8 so you should do a PR on `master` now, not `next` (unless it has breaking changes)

**Comment by Keats** [comment]
I don't see easy wins now so closing this. Perf improvements still welcome though
